assembl/views/search/views.py

```diff
@@ -0,0 +1,62 @@
+import json
+from pyramid.events import NewRequest
+from pyramid.view import view_config
+
+from assembl.indexing.utils import connect
+from assembl.indexing.settings import get_index_settings
+from assembl.indexing.changes import changes, in_tests
+from assembl import models
+from assembl.lib.sqla import get_session_maker
+
+def get_curl_query(query):
+    return "curl -XGET 'localhost:9200/_search?pretty' -d '{}'".format(
+        json.dumps(query).replace("'", "\\u0027"))
+
+
+@view_config(route_name='search', renderer='json')
+def search_endpoint(context, request):
+    es = connect()
+    index_name = get_index_settings()['index_name']
+    query = request.json_body
+#    print get_curl_query(query)
+    result = es.search(index=index_name, body=query)
+
+    # add creator_name in each hit
+    creator_ids = set([hit['_source']['creator_id']
+                       for hit in result['hits']['hits']
+                       if hit['_source'].get('creator_id', None) is not None])
+    session = get_session_maker()
+    creators = session.query(models.AgentProfile.id, models.AgentProfile.name
+        ).filter(models.AgentProfile.id.in_(creator_ids)).all()
+    creators_by_id = dict(creators)
+    # u'query': {u'bool': {u'filter': [{u'term': {u'discussion_id': u'23'}}]}}
+    filters = [fil for fil in query['query']['bool']['filter']]
+    discussion_id = [f.values()[0].values()[0]
```

Comment:
I tested some crafted query, for example having a discussion_id in the hasChild filter that is different from the one in the query, I didn't get any result. Indeed, the intersection of the main query and the hasChild filter is an empty set. And if you don't specify a discussion_id in a hasChild filter, all you will get is a wrong score (number of posts of an user).
The only way you can do a query and have results is by specifying a discussion_id in a bool query. Any query that don't have that will get you an HTTPUnauthorized exception.

After some thoughts, there is maybe a crafted query based on hasChild or hasParent with inner_hits where we can leak info. I'll get back to you on that. The solution would be to remove any inner_hits key from the query before doing the search.

As I understand the security model, you have global roles and only local roles on a discussion. Here we are indexing only PUBLISHED posts, and unindex them if their state change. Even if the posts are published, you may not have the permission to read them because you don't have the read permission on the discussion. The code here takes care of verifying you have the read permission on the discussion, and raise HTTPUnauthorized otherwise.
For the moment, the READ permission on a post is the same as the READ permission of the discussion, so checking the READ permission on the discussion is enough. But this may change in the future. Checking the READ permission for each result may be a good idea. I already retrieve the object for an idea or an agent profile to get the counters, so verifying additional permissions shouldn't be much overhead. We only have 5 results here.

The security modules proposed by X-Pack is a paid service, you need to buy a license. As far I understand it's to protect the elasticsearch rest api if you expose it externally, and the ACLs is I suppose to give the permissions of some endpoints and HTTP verbs to specific users.
We don't need this. We don't expose elasticsearch externally. We only expose it via the pyramid view and you can only do GET localhost:9200/_search here because we do the query via the elasticsearch python binding es.search.